### ISSUE TITLE
Optimize `AutodiffCostFunction` to avoid dense gradients computation for Lie groups

### ISSUE THREAD

**Issue by fantaosha** [facebookresearch/theseus#77]
As discussed offline,  for `AutodiffCostFunction` we actually want sparse jacobians for efficiency, but we'll need to look into using `vmap` to replace the current implementation using `torch.autograd.jacobian` (or some other feasible alternative). This is not an issue when doing `loss.backward()`.

_Originally posted by @luisenp in https://github.com/facebookresearch/theseus/pull/73#discussion_r803055352_

**Comment by fantaosha** [comment]
See #98 and #99.
